File: Servlets/edu/rice/rubbos/servlets/StoreStory.java
package edu.rice.rubbos.servlets;

import edu.rice.rubbos.*;
import java.io.*;
import javax.servlet.*;
import javax.servlet.http.*;
import java.sql.*;

/** Builds the html page with the list of all categories and provides links to browse all
    items in a category or items in a category for a given region */
public class StoreStory extends RubbosHttpServlet
{
  private ServletPrinter sp = null;
  private PreparedStatement stmt = null, stmt2 = null;
  private Connection conn = null;

  public int getPoolSize()
  {
    return Config.BrowseCategoriesPoolSize;
  }

  private void closeConnection()
  {
    try 
    {
      if (stmt != null) stmt.close();	// close statement
    } 
    catch (Exception ignore) 
    {
    }
  }





  /** Build the html page for the response */
  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
  {
    
    sp = new ServletPrinter(response, "StoreStory");
    

        
    String categoryName, nickname, title, body, category, table;
    String password= null;
    int  userId, access;
    ResultSet rs = null, rs2 = null;
    
    nickname = request.getParameter("nickname");
    password = request.getParameter("password");
    title = request.getParameter("title");
    body = request.getParameter("body");
    category = request.getParameter("category");
    
    conn = getConnection();

    if (title == null)
      {
         sp.printHTML("You must provide a story title!<br>");
         return;
      }

    if (body == null)
      {
	 sp.printHTML("<h3>You must provide a story body!<br></h3>");
         return;
      }

    if (category == null)
      {
	 sp.printHTML("<h3>You must provide a category!<br></h3>");
         return;
      }

    sp.printHTMLheader("RUBBoS: Story submission result");

    sp.printHTML("<center><h2>Story submission result:</h2></center><p>\n");
    
    //Authenticate the user
    userId = 0;
    access = 0;

    if ((nickname != null) && (password != null))
	{
	    try
		{
		    stmt = conn.prepareStatement("SELECT id,access FROM users WHERE nickname=\""+ nickname + "\" AND password=\"" + password + "\"");
		    rs = stmt.executeQuery();
		}
	    catch (Exception e)
		{
		    sp.printHTML("ERROR: Authentification query failed" +e);
		    closeConnection();
		    return;
		}
	    try 
		{
		    if (rs.first()) 
			{
			    userId = rs.getInt("id");
			    access = rs.getInt("access");
			}
	      
		} 
	    catch (Exception e) 
		{
		    sp.printHTML("Exception storing story " + e +"<br>");
		    closeConnection();
		}
	}


    table = "submissions";
    if (userId == 0)
      sp.printHTML("Story stored by the 'Anonymous Coward'<br>\n");
    else
    {
      if (access == 0)
        sp.printHTML("Story submitted by regular user " + userId+ "<br>\n");
      else
      {
        sp.printHTML("Story posted by author " +userId+ "<br>\n");
        table = "stories";
      }
    }

    // Add story to database

    
    
       
      try
      {
	  stmt2 = conn.prepareStatement("INSERT INTO "+ table + " VALUES (NULL, \"" + title + "\", \""+ body + "\", NOW(), \"" +  userId + "\", "+ category+")");
        
        rs2 = stmt2.executeQuery();
        if (!rs.first())
        {
          sp.printHTML(" ERROR: Failed to insert new story in database.");
          closeConnection();
          return ;
        }
           }
      catch (SQLException e)
      {
        sp.printHTML("Failed to execute Query for StoreStory: " +e);
        closeConnection();
        return;
      }
    
    
    sp.printHTML("Your story has been successfully stored in the " + table + " database table<br>\n");

    sp.printHTMLfooter();

  }

  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
  {
    doGet(request, response);
  }

}

File: PHP/StoreStory.php
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <body>
    <?php
    $scriptName = "SubmitStory.php";
    include("PHPprinter.php");
    $startTime = getMicroTime();

    $nickname = getSessionPostGetParam('nickname');

    $password = getSessionPostGetParam('password');

    $title = getSessionPostGetParam('title');
    if (!isset($title))
	{
      printError($scriptName, $startTime, "SubmitStory", "You must provide a story title!");
      exit();
    }

    $body = getSessionPostGetParam('body');
    if (!isset($body))
	{
      printError($scriptName, $startTime, "SubmitStory", "You must provide a story body!");
      exit();
    }
      
    $category = getSessionPostGetParam('category');
    if (!isset($category))
	{
      printError($scriptName, $startTime, "SubmitStory", "You must provide a category!");
      exit();
    }

    getDatabaseLink($link);

    printHTMLheader("RUBBoS: Story submission result");

    print("<center><h2>Story submission result:</h2></center><p>\n");

    // Authenticate the user
    $userId = 0;
    $access = 0;
    if (!is_null($nickname) && !is_null($password))
    {
      $result = mysql_query("SELECT id,access FROM users WHERE nickname=\"$nickname\" AND password=\"$password\"", $link);
	  if (!$result)
	  {
		error_log("[".__FILE__."] Authentication query 'SELECT id,access FROM users WHERE nickname=\"$nickname\" AND password=\"$password\"' failed: " . mysql_error($link));
		die("ERROR: Authentication query failed for nickname '$nickname': " . mysql_error($link));
	  }
      if (mysql_num_rows($result) != 0)
      {
        $row = mysql_fetch_array($result);
        $userId = $row["id"];
        $access = $row["access"];
      }
      mysql_free_result($result);
    }

    $table = "submissions";
    if ($userId == 0)
      print("Story stored by the 'Anonymous Coward'<br>\n");
    else
    {
      if ($access == 0)
        print("Story submitted by regular user #$userId<br>\n");
      else
      {
        print("Story posted by author #$userId<br>\n");
        $table = "stories";
      }
    }

    // Add story to database
    $now = date("Y:m:d H:i:s");
    $result = mysql_query("INSERT INTO $table VALUES (NULL, \"$title\", \"$body\", '$now', $userId, $category)", $link);
	if (!$result)
	{
		error_log("[".__FILE__."] Failed to insert new story in database 'INSERT INTO $table VALUES (NULL, \"$title\", \"$body\", '$now', $userId, $category)': " . mysql_error($link));
		die("ERROR: Failed to insert new story in database for user '$userId' and category '$category': " . mysql_error($link));
	}

    print("Your story has been successfully stored in the $table database table<br>\n");
    
    mysql_close($link);
    
    printHTMLfooter($scriptName, $startTime);
    ?>
  </body>
</html>
